### PR TITLE
Fix file descriptor type to be `File` and not `Socket` for async

### PR DESCRIPTION
**Pull request by andersfugmann**
This fixes a hang bug on s390x, as the file descriptor created by ZMQ is in fact a file and not a socket (verified by manual adding code to call `Fd.Kind.infer_using_stat`). 

The error was spotted on s390x release tests, where the tests would hang. Once merged I will move the tag for 5.2.0 and recreate a release. See https://github.com/ocaml/opam-repository/pull/22382

### CODE REVIEW

**Comment by Leonidas-from-XIV** [comment]
Good debugging! I'll take a look at the PR hopefully tomorrow but I suggest not moving the tag, just create 5.2.1 and leave 5.2.0 unsubmitted to OPAM. There is not much to be gained from reusing version numbers but it can create confusion.

**Comment by andersfugmann** [comment]
Good point. I will release a 5.2.1 instead once this is merged and close the current opam-repository PR.

**Comment by andersfugmann** [comment]
Just want to mention that I have tested this locally in a z390x docker image and verified that this PR indeed fixes the problem. 
Its not possible to change the logic to automatically determine the type of file descriptor returned by zmq without introducing breaking changes to the API, as `Fd.Kind.infer_using_stat` returns a deferred.

File: zmq-async/src/deferred.ml [hunk 1]
```diff
@@ -35,7 +35,7 @@ module Fd = struct
   type 'a t' = 'a t
   type t = Fd.t
   let create fd =
-    Fd.create (Fd.Kind.Socket `Active) fd (Base.Info.of_string "<zmq>")
+    Fd.create Fd.Kind.File fd (Base.Info.of_string "<zmq>")
```
Comment (Leonidas-from-XIV) [inline zmq-async/src/deferred.ml:38]: So if I understand correctly, an `Fd.t` is created and the error is that here we specify a socket but in reality it is a `file` and this difference is only relevant on `s390x` because presumably other platforms don't particularly care?

My second comment was exactly "can we determine this automatically instead of hardcoding" which is preempted by the fact that this is an API change. Annoying, but I understand the rationale. An option we might consider is to use `Unix.fstat` but I assume this is a blocking operation (and part of `Unix` which would break `Unix`-less platforms) so that can't be done sensibly?

**Review by Leonidas-from-XIV — APPROVED** [review]
I have a comment to confirm my understanding of the bug and the solution, but if I understood it correctly this looks like a good fix - at least as good as it gets without breaking the API.

File: zmq-async/src/deferred.ml (hunk 1 above)
Comment (andersfugmann) [reply zmq-async/src/deferred.ml:38]: 1. Yes. It is a file determined by calling `Fd.infer_using_stat`, and apparently the only platform who cares is the s390x which is a big endian platform. One difference I can spot, is that if the fd is a `socket`, async will try to set the fd in non blocking mode - but not files as linux will ignore that anyways. Why this only has an effect on big endian systems is beyond me.

2. We could call `Unix.fstat` to determine if the socket supports non-blocking, but I don't really think it matters. The type is used mainly to control setting blocking vs non-blocking options on the file descriptor. We never reads or writes to the file descriptor, but just read the state (if reads or writes _would_ block), so setting non_blocking is a noop seen from our use case.  But... The zmq documentation indicates that the file descriptor returned is used internally also so modifying it seems a bad idea. 

Digging further into this, `Fd.create` has an option `?avoid_nonblock_if_possible:bool` (`async_unix.0.14`) (and `?avoid_setting_nonblock:bool` for `async_unix.0.15`). Setting this to true keeping the kind as `Fd.Kind.Socket 'Active`, also fixes the problem. When not set, the test spins using 100% cpu forever.

If only janestreet would keep the API stable, we could have forced never setting the non-blocking mode. For now we need to rely on the fact that non-blocking is never set on files. 

In conclusion, I don't think we need to care too much about the actual type, but just make sure that we never change the fd to non blocking.
